### ISSUE TITLE
Composable routes

### ISSUE THREAD

**Issue by emanuelkoczwara** [elm-community/elm-route#2]
Hello,

Now I have something like this:

```
articleListR   = ArticleListPage   := "articles" <//> static ""
articleDetailR = ArticleDetilPage  := "articles" <//> int
articleCreateR = ArticleCreatePage := "articles" <//> static "create"
articleUpdateR = ArticleUpdatePage := "articles" <//> int <> "update"
articleDeleteR = ArticleDeletePage := "articles" <//> int <> "delete"
```

If I add users, groups etc. to my app, you can guess how it would look like. My `Page` type looks like this:

```
type Page
  = ArticleListPage ()
  | ArticleDetilPage Int
  | ArticleCreatePage ()
  | ArticleUpdatePage Int
  | ArticleDeletePage Int
```

And here the same story, this kind of flat sitemap is going to grow fast in real (big) application. I would like to define `listR`, `detailR`, `createR`, `updateR` and `deleteR` like this:

```
listR   = ListPage   := static ""
detailR = DetailPage := int
createR = CreatePage := static "create"
updateR = UpdatePage := int <> "update"
deleteR = DeletePage := int <> "delete"
```

and then reuse them. I would like to define one route for each submodule of my app like:

```
articleR = ArticlePages  := "articles"
userR    = UserPages     := "users"
groupR   = GroupPages    := "groups"
```

and the I would like to combine them as I need, for example:

```
route (articleR <//> listR) []   ==> "/articles/"
route (userR <//> updateR) [1]   ==> "/users/1/update/"
```

My new `Page` type would look something like this:

```
type BasicPage
  = ListPage ()
  | DetailPage Int
  | CreatePage ()
  | UpdatePage Int
  | DeletePage Int

type Page
  = ArticlePage BasicPage
  | UserPage BasicPage
  | GroupPage BasicPage
```

Is it possible to do something like this with current version? And if yes, then how?

**Comment by Bogdanp** [comment]
That's not possible right now but something like this would be possible with a few small changes to the library:

``` elm
module Reuse where

import Route exposing (..)

type Operation
  = List ()
  | Create ()
  | Read Int
  | Update Int
  | Delete Int

listR = List := static ""
createR = Create := static "create"
readR = Read := int
updateR = Update := int <> "update"
deleteR = Delete := int <> "delete"
operationRouter = router [ listR, createR, readR, updateR, deleteR ]

renderOperation : Operation -> String
renderOperation o =
  case o of
    List () -> Route.reverse listR []
    Create () -> Route.reverse createR []
    Read id -> Route.reverse readR [toString id]
    Update id -> Route.reverse updateR [toString id]
    Delete id -> Route.reverse deleteR [toString id]

type Resource
  = Article Operation
  | Book Operation

articleR = "articles" <//> child Article operationRouter
bookR = "books" <//> child Book operationRouter
resourceRouter = router [ articleR , bookR ]

match : String -> Maybe Resource
match =
  Route.match resourceRouter

render : Resource -> String
render r =
  case r of
    Article op -> Route.reverse articleR [] ++ renderOperation op
    Book op -> Route.reverse bookR [] ++ renderOperation op
```

It would have the unfortunate side-effect of allowing things like:

``` elm
> match (router [int]) "/1"
Just 1
```

But I don't think that's too bad and existing code would continue working with these changes. That being said, I'll think it through a little more thoroughly tomorrow and if I can't come up with any real downside then I'll release a version that allows this.

**Comment by Bogdanp** [comment]
Alright, I've published version `2.0.0` with changes that make the above possible.
